liesel/band.py

```diff
@@ -0,0 +1,232 @@
+from functools import partial
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+import scipy.sparse as sps
+
+
+# @jax.jit not useful, f is compiled inside scan anyway, this makes it possible to return int.
+def bandwidth(full) -> int:
+    """
+    Returns the bandwidth p of `full` as an integer.
+
+    ## Parameters
+
+    - `full`: A full (n x n) matrix.
+    """
+
+    def f(j, row):  # Iterate over Row
+        def g(i, val):  # Iterate over value in Row
+            return i + 1, jax.lax.cond(val != 0, jnp.abs, lambda _: -1, i - j)
+
+        _, L = jax.lax.scan(g, 0, row)
+        return j + 1, L
+
+    _, b = jax.lax.scan(f, 0, full)
+    return int(jnp.max(b))
+
+
+@partial(jax.jit, static_argnames="bw")
+def to_band(full, bw, lower=True):
+    """
+    ## Parameters
+
+    - `full`: A full (n x n) matrix.
+    - `bw`: The bandwidth p.
+    - `lower`: Whether to return the (p x n) band in lower-diagonal ordered form.
+    """
+    n = full.shape[0]
+    bw += 1
+
+    # We only access the upper triangular. If we get a Matrix in lower-triangular (L) form, we convert.
+    full = jnp.where(jnp.array_equal(full, jnp.tril(full)), full.T, full)
+
+    # TODO: lax.fori_loop here?
+
+    def lo():
+        Ab = jnp.zeros((bw, n))
+        for k in range(bw):
+            Ab = Ab.at[k, :(n - k)].set(jnp.diag(full, k))
+        return Ab
+
+    def up():
+        Ab = jnp.zeros((bw, n))
+        for k in range(bw):
+            Ab = Ab.at[k, k:].set(jnp.diag(full, k))
+        return jnp.flipud(Ab)
+
+    return jax.lax.cond(lower, lo, up)
+
+
+# no @jax.jit
+def permute(full, lower=True):
+    """
+    Permutes `full` with the reverse Cuthill-McKee algorithm and returns the result
+    as a (p x n) band.
+
+    ## Parameters
+
+    - `full`: A full, symmetric(!) (n x n) matrix.
+    - `lower`: Whether to return the (p x n) band in lower-diagonal ordered form.
+    """
+
+    if not np.allclose(full, full.T):
+        raise ValueError("Argument `full` is not symmetric.")
+
+    p = sps.csgraph.reverse_cuthill_mckee(sps.csr_matrix(full), symmetric_mode=True)
+    P = np.vstack([np.eye(1, full.shape[0], k) for k in p])
+    fullP = P @ full @ P.T
```

Comment:
Can we compute p without constructing `fullP` and using `bandwidth` on it?

Reply:
I don't think so. `full` can be made banded sooner, but it is hard to apply the Permutation Matrix in a banded format.
Bandwidth also can't be extracted using $p$

Reply:
That's okay. By the way, thinking about this function again, I would propose to just return `fullP` and the permutation. Then the user can decide if they want to convert the result to a band matrix, which might not always be the case.

Also, could you permute `full` by indexing rather than by matrix multiplication? That should save us some memory.

Reply:
All done, maybe also return indicies and not $p$?